staking/integration-tests/tests/stability.rs

```diff
@@ -0,0 +1,439 @@
+        signer::Signer,
+    },
+    staking::{
+        error::ErrorCode as StakingErrorCode,
+        state::positions::{
+            Position,
+            PositionState,
+            Target,
+            TargetWithParameters,
+        },
+    },
+    std::{
+        cmp::min,
+        convert::TryInto,
+        fmt::Debug,
+    },
+};
+
+// These constants define the parameters of the stability test.
+const MAX_DELEGATION_AMOUNT: u64 = 2 * FRAC_64_MULTIPLIER;
+const MAX_UNDELEGATION_AMOUNT: u64 = FRAC_64_MULTIPLIER / 10;
+const NUM_DELEGATORS: u64 = 1_000;
+const NUM_OPERATIONS: u64 = 10_000;
+const MAX_CAP: u64 = 100 * FRAC_64_MULTIPLIER;
+
+#[derive(Debug)]
+struct KeypairWrapper(Keypair);
+
+impl Clone for KeypairWrapper {
+    fn clone(&self) -> Self {
+        KeypairWrapper(self.0.insecure_clone())
+    }
+}
+
+#[derive(Debug, Clone)]
+enum Operation {
+    Delegate {
+        publisher: usize,
+        delegator: usize,
+        amount:    u64,
+    },
+    Undelegate {
+        publisher: usize,
+        delegator: usize,
+        amount:    u64,
+    },
+    AdvanceDelegationRecord {
+        publisher: usize,
+        delegator: usize,
+    },
+    Advance {
+        cap: u64,
+    },
+    Slash {
+        publisher: usize,
+        amount:    u64,
+    },
+}
+
+#[derive(Clone)]
+struct StabilityTestProps {
+    operations: Vec<Operation>,
+    publishers: Vec<KeypairWrapper>,
+    delegators: Vec<KeypairWrapper>,
+}
+
+impl Debug for StabilityTestProps {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("StabilityTestProps")
+    }
+}
+
+impl Arbitrary for StabilityTestProps {
+    fn arbitrary(g: &mut Gen) -> Self {
+        let num_publishers = usize::arbitrary(g) % (MAX_PUBLISHERS - 10) + 10;
+
+        // TODO: make stable
+        let publishers: Vec<KeypairWrapper> = (0..num_publishers)
+            .map(|_| KeypairWrapper(Keypair::new()))
+            .collect();
+
+        let delegators: Vec<KeypairWrapper> = (0..NUM_DELEGATORS)
+            .map(|_| KeypairWrapper(Keypair::new()))
+            .collect();
+
+        let operations = (0..NUM_OPERATIONS)
+            .map(|_| get_random_operation(g, &publishers, &delegators))
+            .collect();
+
+        StabilityTestProps {
+            operations,
+            publishers,
+            delegators,
+        }
+    }
+}
+
+fn get_random_operation(
+    g: &mut Gen,
+    publishers: &[KeypairWrapper],
+    delegators: &[KeypairWrapper],
+) -> Operation {
+    let delegate_weight = 500;
+    let undelegate_weight = 250;
+    let advance_delegation_record_weight = 250;
+    let advance_weight = 10;
+    let slash_weight = 1;
+    let total_weight = delegate_weight
+        + undelegate_weight
+        + advance_delegation_record_weight
+        + advance_weight
+        + slash_weight;
+
+    let rnd = u64::arbitrary(g) % total_weight;
+
+    if rnd < delegate_weight {
+        Operation::Delegate {
```

Comment:
mmm could we throw some governance create/close_positions too?

Reply:
Done